net: Fix solicited-node multicast group handling for IPv6 address resolution

pfalcon opened the pull request:
~~~

net: if: Join solicited-node multicast addr for each unicast addr  …
https://tools.ietf.org/html/rfc4862#section-5.4.2 :

"""
Before sending a Neighbor Solicitation, an interface MUST join the
all-nodes multicast address and the solicited-node multicast address
of the tentative address.
"""

So, joining should happen before sending DAD packets, and it should
happen for each unicast address added. This is achieved by joining
from net_if_ipv6_addr_add() call. Note that we already leave
solicited-node group from net_if_ipv6_addr_rm(). In particular, we
leave it if DAD fails (as that function is called in this case).

Fixes #5282.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>
~~~

pfalcon committed:
net: if: Join solicited-node multicast addr for each unicast addr

https://tools.ietf.org/html/rfc4862#section-5.4.2 :

"""
Before sending a Neighbor Solicitation, an interface MUST join the
all-nodes multicast address and the solicited-node multicast address
of the tentative address.
"""

So, joining should happen before sending DAD packets, and it should
happen for each unicast address added. This is achieved by joining
from net_if_ipv6_addr_add() call. Note that we already leave
solicited-node group from net_if_ipv6_addr_rm(). In particular, we
leave it if DAD fails (as that function is called in this case).

Fixes #5282.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
drivers: eth_mcux: Disable promiscuous mode by default

Now that proper solicited-node multicast group joing is implemented,
promiscuous mode's purpose is reduced to just debugging needs.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Also includes patch to finally disabled promisc mode on frdm_k64f, so adding @MaureenHelm and @dleach02 to reviewers.

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -707,6 +696,17 @@ struct net_if_addr *net_if_ipv6_addr_add(struct net_if *iface,
+		 * MUST join the all-nodes multicast address and the
+		 * solicited-node multicast address of the tentative address."
+		 */
+		/* The allnodes multicast group is only joined once as
```
you can merge the 2 comments

pfalcon replied:
In principle yes, but those are two *different* comments, of different levels. To emphasize the difference, it's better to treat them as "annotations". First is normative reference annotation, applying to whole block of code (2 lines). 2nd is behavior-descriptive annotation, applying to just single statement. Merging two doesn't make much sense, and could lead to decreased clarity of plain confusion.

tbursztyka replied:
true

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -461,17 +461,6 @@ static void dad_timeout(struct k_work *work)
-		 */
-		join_mcast_allnodes(iface);
-
-		join_mcast_solicit_node(iface, &ifaddr->address.in6_addr);
```
Sounds fine to me, but maybe you'll need to address the case when DAD is a failure: what to do with the mcast then? (or is it relevant actually).
Before, because it was joining mcast only on success, failure case did not have to handle any of it, but now it's different.

pfalcon replied:
That's addressed in the commit message:

> This is achieved by joining
from net_if_ipv6_addr_add() call. Note that we already leave
solicited-node group from net_if_ipv6_addr_rm(). In particular, we
leave it if DAD fails (as that function is called in this case).

tbursztyka replied:
indeed line 745 yes

pfalcon committed:
tests: net: mld: Update after solicited-node mcast refactor.

1. MLD events now can happen before the test starts to run, (at iface
initialization time), so use static initialization of test semaphore.
2. Don't use a well-known multicast group like ff02::1 (all nodes),
as they are managed by the underlying stack. Use a dedicated test
address (ff10::1 here).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Now the build passes. @tbursztyka, please re-review.
